knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+        if "name" in dict:
+            yield dict["name"]
+
+
+def create_json_list_mapping(
+    path_to_json: str,
+    list_name: str,
+    language_label: str
+) -> dict[str, str]:
+    """
+    Often, data sources contain list values named after the "label" of the JSON project list node, instead of the "name"
+    which is needed for the `dsp-tools xmlupload`. In order to create a correct XML, you need a dictionary that maps the
+    "labels" to their correct "names".
+
+    Alternatively, consider using the method create_json_excel_list_mapping(), which also creates a dictionary, but maps
+    values from your data source to list node names from the JSON project file, based on similarity.
+
+    Args:
+        path_to_json: path to a JSON project file (a.k.a. ontology)
+        list_name: name of a list in the JSON project (works also for nested lists)
+        language_label: which language of the label to choose
+
+    Returns:
+        a dictionary of the form {label: name}
+    """
+    with open(path_to_json) as f:
+        json_file = json.load(f)
+    json_subset = list()
+    for numbered_json_obj in json_file["project"]["lists"]:
+        if numbered_json_obj["name"] == list_name:
+            json_subset.append(numbered_json_obj)
+    # json_subset is a list containing one item, namely the json object containing the entire json-list
+
+    res = {}
+    for label, name in _name_label_mapper_iterator(json_subset, language_label):
+        if name != list_name:
+            res[label] = name
+            res[label.strip().lower()] = name
+
+    return res
+
+
+def _name_label_mapper_iterator(json_subset: list[dict[str, Any]], language_label: str) -> Iterable[tuple[str, str]]:
+    """
+    returns (label, name) pairs of JSON project list entries
+    """
+    for node in json_subset:
+        # node is the json object containing the entire json-list
+        if "nodes" in node:
+            # "nodes" is the json sub-object containing the entries of the json-list
+            for value in _name_label_mapper_iterator(node["nodes"], language_label):
+                yield value
+                # "value" is a (label, name) pair of a single list entry
+        if "name" in node:
+            yield (node["labels"][language_label], node["name"])
+            # the actual values of the name and the label
+
+
+def write_xml(root: etree._Element, filepath: str) -> None:
+    """
+    Write the finished XML to a file
+
+    Args:
+        root: etree Element with the entire XML document
+        filepath: where to save the file
+
+    Returns:
+        None
+    """
+    etree.indent(root, space="    ")
+    xml_string = etree.tostring(root, encoding="unicode", pretty_print=True)
+    xml_string = '<?xml version="1.0" encoding="UTF-8"?>\n' + xml_string
+    xml_string = xml_string.replace("&lt;", "<")
+    xml_string = xml_string.replace("&gt;", ">")
+    with open(filepath, "w", encoding="utf-8") as f:
+        f.write(xml_string)
```

Comment:
why those string operations? to get the processing instructions in there?

Reply:
I'd have to look up how, but it can be solved directly in lxml so that you don't need to do any string juggling

Reply:
I agree, and I read https://lxml.de/api/lxml.etree-module.html#tostring, but I don't suceed.

Reply:
I just had a quick look: When Lina faced the same issue in the eHKKA project, we used an `ElementTree` object, which can hold the processing instruction, and on which one can call `write()`.  
Here's some code I copied from there:

```python
    tree = et.ElementTree(root) # root is an `et._Element`
    # add processing instruction to the XML document
    pi = et.ProcessingInstruction('xml-model',
                                  'href="https://tei-c.org/release/xml/tei/custom/schema/xsd/tei_all.xsd" ' +
                                  'type="application/xml" ' +
                                  'schematypens="http://www.w3.org/2001/XMLSchema"')
    root.addprevious(pi)
    # write the XML data to the file
    tree.write(file_path, pretty_print=True, xml_declaration=True, encoding="UTF-8")
```

looking at it now, you don't even need the processing instruction, so you can ignore that part. But I think the relevant bits are `xml_declaration=True, encoding="UTF-8"` (which you could also do in `tostring()`). But I guess it's still more elegant to leave all the work to the ElementTree object.

So the function could literally be no more than:
```python
def write_xml(root: etree._Element, filepath: str) -> None:
    tree = et.ElementTree(root)
    tree.write(file_path, pretty_print=True, xml_declaration=True, encoding="UTF-8")
```
And that might just work ok.